### ISSUE TITLE
Regarding marginalization of the Gaussian noise

### ISSUE THREAD

**Issue by jamshids** [tanyanair/segmentation_uncertainty#3]
Hi Tanya,

I think there is a mismatch between what your code computes and what [1] is proposing (reference [8] in your MICCAI paper). In [1], they are proposing to marginalize out Gaussian distribution over the output logits (equation (11)) by taking the log of the expected values of softmax probabilities, i.e. logE[softmax_y(f(x) + eps\*sigma(x))] (where y is the observed class of x). On the other hand, your function `bunet.utils.weighted_mc_xentropy` seems to reverse the operations log and expectation, i.e. it computes E[ log(softmax_y(f(x)+eps*sigma(x))) ] instead. That's because first `tf.nn.weighted_cross_entropy_with_logits` is applied on the noisy logits, which includes taking the logarithm, and then the results are averaged. I am not sure how much these values would be different numerically, and how much this difference effects the training process. 

Please let me know if I am misunderstanding your codes.

Best,
Jamshid

[1] Kendall, A., Gal, Y.: What uncertainties do we need in bayesian deep learning for computer vision? NIPS pp. 5580–5590 (2017)
